experiments/yarn/src/main/scala/io/gearpump/experiments/yarn/appmaster/YarnAppMaster.scala

```diff
@@ -0,0 +1,357 @@
+      }
+    case ShutdownApplication =>
+      LOG.error("ShutdownApplication")
+      nmClient.stop()
+      rmClient.shutdownApplication()
+    case ResourceManagerException(ex) =>
+      LOG.error("ResourceManagerException: " + ex.getMessage, ex)
+      nmClient.stop()
+      rmClient.failApplication(ex)
+    case Kill =>
+      LOG.info("Kill: User asked to shutdown the application")
+      sender ! CommandResult(success = true)
+      self ! ShutdownApplication
+  }
+
+  private def service(config: Config, masters: List[MasterInfo], workers: List[WorkerInfo]): Receive  = box {
+    case GetActiveConfig(clientHost) =>
+      LOG.info("GetActiveConfig: Get active configuration for client: " + clientHost)
+      val filtered = ClusterConfig.filterOutDefaultConfig(config.withValue(Constants.GEARPUMP_HOSTNAME,
+        ConfigValueFactory.fromAnyRef(clientHost)))
+      sender ! ActiveConfig(filtered)
+    case QueryVersion =>
+      LOG.info("QueryVersion")
+      sender ! Version(Util.version)
+    case QueryClusterInfo =>
+      LOG.info("QueryClusterInfo")
+      val masterContainers = masters.map{master =>
+        master.id.toString + s"(${master.nodeId.toString})"
+      }
+
+      val workerContainers = workers.map{worker=>
+        worker.id.toString + s"(${worker.nodeId.toString})"
+      }
+      sender ! ClusterInfo(masterContainers, workerContainers)
+    case AddWorker(count) =>
+      if (count == 0) {
+        sender ! CommandResult(success = true)
+      } else {
+        LOG.info("AddWorker: Start to add new workers, count: " + count)
+        workerCount += count
+        requestWorkerContainers(count)
+        context.become(startingWorkers(count, masters, workers))
+        sender ! CommandResult(success = true)
+      }
+    case RemoveWorker(worker) =>
+      val workerId = ContainerId.fromString(worker)
+      LOG.info(s"RemoveWorker: remove worker $workerId")
+      val info = workers.find(_.id.toString == workerId.toString)
+      if (info.isDefined) {
+        nmClient.stopContainer(info.get.id, info.get.nodeId)
+        sender ! CommandResult(success = true)
+        val remainWorkers = workers.filter(_.id != info.get.id)
+        context.become(service(config, masters, remainWorkers))
+      } else {
+        sender ! CommandResult(success = false, "failed to find worker " + worker)
+      }
+  }
+
+  private def unHandled: Receive = {
+    case other =>
+      LOG.info(s"Received unknown message $other")
+  }
+
+  private def requestMasterContainers(masters: Int) = {
+    LOG.info(s"Request resource for masters($masters)")
+    val containers = (1 to masters).map(
+      i => Resource.newInstance(masterMemory, masterVCores)
+    ).toList
+    rmClient.requestContainers(containers)
+  }
+
+  private def launchMaster(container: Container): HostPort = {
+    LOG.info(s"Launch Master on container " + container.getNodeHttpAddress)
+    val host = container.getNodeId.getHost
+
+    val port = Util.findFreePort.get
```

Comment:
the free port is on the YarnAppMaster's host, not on the allocated container's, right?